src/cirrus/lib/task.py

```diff
@@ -0,0 +1,208 @@
+    def parameters(self):
+        return self.process_definition['tasks'].get(self._name, {})
+
+    @property
+    def output_options(self):
+        return self.process_definition.get('output_options', {})
+
+    def validate(self):
+        # put validation logic on input Items and process definition here
+        pass
+
+    def download_assets(self, assets: Optional[List[str]]=None):
+        """Download provided asset keys for all items in payload. Assets are saved in workdir in a
+           directory named by the Item ID
+
+        Args:
+            assets (Optional[List[str]], optional): List of asset keys to download. Defaults to all assets.
+        """
+        for i, item in enumerate(self.items):
+            outdir = op.join(self._workdir, item['id'])
+            makedirs(outdir, exist_ok=True)
+            self.items[i] = download_item_assets(item, path=outdir, assets=assets)
+
+
+    def upload_assets(self, assets: Optional[List[str]]=None):
+        if self._local:
+            self.logger.warn('Running in local mode, assets not uploaded')
+            return
+        for i, item in enumerate(self.items):
+            self.items[i] = upload_item_assets(item, **self.output_options)
+
+    # this should be in PySTAC
+    @classmethod
+    def create_item_from_item(self, item):
+        # create a derived output item
+        
+            links = [l['href'] for l in item['links'] if l['rel'] == 'self']
+            if len(links) == 1:
+                # add derived from link
+                item ['links'].append({
+                    'title': 'Source STAC Item',
+                    'rel': 'derived_from',
+                    'href': links[0],
+                    'type': 'application/json'
+                })
+
+    @abstractmethod
+    def process(self):
+        """Main task logic - virtua
+
+        Returns:
+            [type]: [description]
+        """
+        # download assets of interest, this will update self.items
+        #self.download_assets(['key1', 'key2'])
+        # do some stuff
+        #self.upload_assets(['key1', 'key2'])
+        pass
+
+    @property
+    def output_payload(self) -> Dict:
+        
+        # assemble return payload
+        payload = self.remaining_payload
+        payload.update({
+            'features': self.items,
+            'process': self.process_definition
+        })
+        return payload
+
+    @classmethod
+    def handler(cls, payload, **kwargs):
+        task = cls(payload, **kwargs)
+        try:
+            task.process()
+            return task.output_payload
+        except Exception as err:
+            task.logger.error(err, exc_info=True)
+            raise err
+        finally:
+            # remove work directory if not running locally
+            if task._tmpworkdir:
+                task.logger.debug('Removing work directory %s' % task._workdir)
+                rmtree(task._workdir)
+        
+    @classmethod
+    def parse_args(cls, args):
```

Comment:
For batch specifically, it could be advantageous to allow users to add subcommands/args in the case where they want to have a single container implement several different commands. Thus one image could be used by multiple job definitions. I'm not sure how this would plug in with the `local` vs `cirrus` idea (it seems like mode of operation is separate from which command is being run).

Maybe this is as simple as using `parser.parse_known_args` instead of `parser.parse_args`, and passing any remainder to `process` for the end user to parser. That said, I've never used it but understand it has some edge cases that can be tricky, so I'm not sure if that would work well or not.

Reply:
Refactored the code so that creating the parser is it's own function now that users can fetch then add on it.  They would then redefine the `cli` function.